README.md
<!--
SPDX-FileCopyrightText: Magenta ApS

SPDX-License-Identifier: MPL-2.0
-->

# OS2MO-FastAPI-Utils

Utility library with various reusable FastAPI components.


## Instrumentation Usage
Install into your project using `pip`:
```
pip install os2mo-fastapi-utils
```

Then import it inside a Python file:
```
from fastapi import FastAPI
from os2mo_fastapi_utils.tracing import setup_instrumentation

app = FastAPI()

@app.get("/")
async def root():
    return {"message": "Hello World"}

app = setup_instrumentation(app)
```


## Keycloak Autentication

The `auth` package in this module provides a Keycloak auth `Depends` coroutine
for FastAPI. Do (e.g.) the following to use this in FastAPI:
```
from fastapi import FastAPI, Depends
from structlog import get_logger

from os2mo_fastapi_utils.auth.exceptions import AuthenticationError
from os2mo_fastapi_utils.auth.oidc import get_auth_dependency
from os2mo_fastapi_utils.auth.oidc import get_auth_exception_handler
from myproject.models import MyToken

logger = get_logger()

auth = get_auth_dependency(
           host='keycloak',
           port=8081,
           realm='mo',
           token_url_path='service/token',
           token_model=MyToken
       )

app = FastAPI()
app.add_exception_handler(
    AuthenticationError,
    get_auth_exception_handler(logger)
)

@app.get("/")
async def root(token: Token = Depends(auth)):
    return {"message": "Hello World"}
```


CHANGELOG.md
CHANGELOG
=========

1.3.2 - 2023-05-23
------------------

fix: [#56049] properly log authentication exceptions

1.3.1 - 2023-04-04
------------------

[#55602] pep 561 compliance

1.3.0 - 2023-01-18
------------------

[#54112] Add token expiry leeway

1.2.0 - 2021-12-08
------------------

[#47508] Implement automatic versioning through autopub


tests/test_oidc.py
# SPDX-FileCopyrightText: 2019-2020 Magenta ApS
# SPDX-License-Identifier: MPL-2.0
import asyncio
import unittest.mock
from datetime import datetime
from uuid import UUID
from uuid import uuid4

import jwt
from cryptography.hazmat.primitives import serialization
from fastapi.exceptions import HTTPException
from jwt.exceptions import ExpiredSignatureError
from jwt.exceptions import InvalidAudienceError
from jwt.exceptions import InvalidSignatureError
from jwt.exceptions import InvalidTokenError
from jwt.exceptions import PyJWTError
from pydantic import ValidationError
from starlette.status import HTTP_401_UNAUTHORIZED
from starlette.status import HTTP_500_INTERNAL_SERVER_ERROR
from structlog import get_logger

from os2mo_fastapi_utils.auth.exceptions import AuthenticationError
from os2mo_fastapi_utils.auth.models import RealmAccess
from os2mo_fastapi_utils.auth.models import Token
from os2mo_fastapi_utils.auth.oidc import get_auth_dependency
from os2mo_fastapi_utils.auth.oidc import get_auth_exception_handler


class TestOIDC(unittest.TestCase):
    def setUp(self) -> None:
        # Used for mocking the Keycloak public key
        with open("tests/mocking/auth/jwtRS256.key.pub", "rb") as fp:
            public_key = serialization.load_pem_public_key(fp.read())
        self.signing = jwt.PyJWT({"kty": "RSA"})
        # self.signing.key = public_key
        setattr(self.signing, "key", public_key)

        # Used for mocking the Keycloak private key
        with open("tests/mocking/auth/jwtRS256.key", "rb") as fp:
            self.private_key = fp.read()

        # Used for mocking the OIDC token
        self.parsed_token = {
            "acr": "1",
            "allowed-origins": ["http://localhost:5001"],
            "azp": "mo",
            "email": "bruce@kung.fu",
            "email_verified": False,
            "exp": int(datetime.now().timestamp()) + 300,
            "family_name": "Lee",
            "given_name": "Bruce",
            "iat": int(datetime.now().timestamp()),
            "iss": "http://localhost:8081/auth/realms/mo",
            "jti": "25dbb58d-b3cb-4880-8b51-8b92ada4528a",
            "name": "Bruce Lee",
            "preferred_username": "bruce",
            "scope": "email profile",
            "session_state": "d94f8dc3-d930-49b3-a9dd-9cdc1893b86a",
            "sub": "c420894f-36ba-4cd5-b4f8-1b24bd8c53db",
            "typ": "Bearer",
            "uuid": "99e7b256-7dfa-4ee8-95c6-e3abe82e236a",
        }

        self.auth = get_auth_dependency(
            host="keycloak",
            port=8080,
            realm="mo",
            token_url_path="service/token",
            token_model=Token,
        )

        # For the async tests
        self.loop = asyncio.get_event_loop()

    @staticmethod
    def generate_token(parsed_token: dict, key: str) -> str:
        """
        Generate a request containing an auth header with a OIDC Bearer token
        :param parsed_token: parsed token (see example above)
        :param key: The JWK to sign the token with
        """
        token = jwt.encode(parsed_token, key, algorithm="RS256")
        return token

    def test_auth_exception_handler_return_401_for_client_side_error(self):
        exc = AuthenticationError(InvalidSignatureError())
        authentication_exception_handler = get_auth_exception_handler(get_logger())

        self.assertEqual(
            HTTP_401_UNAUTHORIZED,
            authentication_exception_handler(None, exc).status_code,
        )

    def test_auth_exception_handler_return_500_for_server_side_error(self):
        exc = AuthenticationError(PyJWTError())
        authentication_exception_handler = get_auth_exception_handler(get_logger())

        self.assertEqual(
            HTTP_500_INTERNAL_SERVER_ERROR,
            authentication_exception_handler(None, exc).status_code,
        )

    @unittest.mock.patch(
        "os2mo_fastapi_utils.auth.oidc.jwt.PyJWKClient.get_signing_key_from_jwt"
    )
    def test_auth_decodes_token(self, mock_get_signing_key_from_jwt):
        # Mock the public signing.key used in the auth function
        mock_get_signing_key_from_jwt.side_effect = [self.signing]

        # Create auth request with token signed by correct key
        token = TestOIDC.generate_token(self.parsed_token, self.private_key)

        actual_token = self.loop.run_until_complete(self.auth(token))
        expected_token = Token(
            azp="mo",
            email="bruce@kung.fu",
            preferred_username="bruce",
            uuid=UUID("99e7b256-7dfa-4ee8-95c6-e3abe82e236a"),
        )

        self.assertEqual(expected_token, actual_token)

    @unittest.mock.patch(
        "os2mo_fastapi_utils.auth.oidc.jwt.PyJWKClient.get_signing_key_from_jwt"
    )
    def test_leeway(self, mock_get_signing_key_from_jwt):
        # Mock the public signing.key used in the auth function
        mock_get_signing_key_from_jwt.side_effect = [self.signing]

        # Set "exp" to a slightly expired timestamp
        self.parsed_token["exp"] = int(datetime.now().timestamp()) - 3

        # Create auth request with token signed by correct key
        token = TestOIDC.generate_token(self.parsed_token, self.private_key)

        actual_token = self.loop.run_until_complete(self.auth(token))
        expected_token = Token(
            azp="mo",
            email="bruce@kung.fu",
            preferred_username="bruce",
            uuid=UUID("99e7b256-7dfa-4ee8-95c6-e3abe82e236a"),
        )

        self.assertEqual(expected_token, actual_token)

    @unittest.mock.patch(
        "os2mo_fastapi_utils.auth.oidc.jwt.PyJWKClient.get_signing_key_from_jwt"
    )
    def test_raise_exception_for_invalid_signature(self, mock_get_signing_key_from_jwt):
        # Mock the public signing.key used in the auth function
        mock_get_signing_key_from_jwt.side_effect = [self.signing]

        with open("tests/mocking/auth/hackers-jwtRS256.key", "rb") as fp:
            hackers_key = fp.read()

        # Create auth request with token signed with hackers key
        token = TestOIDC.generate_token(self.parsed_token, hackers_key)

        with self.assertRaises(AuthenticationError) as err:
            self.loop.run_until_complete(self.auth(token))
            self.assertTrue(isinstance(err.exception.exc, InvalidSignatureError))

    @unittest.mock.patch(
        "os2mo_fastapi_utils.auth.oidc.jwt.PyJWKClient.get_signing_key_from_jwt"
    )
    def test_raise_exception_for_expired_token(self, mock_get_signing_key_from_jwt):
        # Mock the public signing.key used in the auth function
        mock_get_signing_key_from_jwt.side_effect = [self.signing]

        # Set "exp" to an expired timestamp
        self.parsed_token["exp"] = int(datetime.now().timestamp()) - 6

        # Create auth request with token signed by correct key
        token = TestOIDC.generate_token(self.parsed_token, self.private_key)

        with self.assertRaises(AuthenticationError) as err:
            self.loop.run_until_complete(self.auth(token))
            self.assertTrue(isinstance(err.exception.exc, ExpiredSignatureError))

    @unittest.mock.patch(
        "os2mo_fastapi_utils.auth.oidc.jwt.PyJWKClient.get_signing_key_from_jwt"
    )
    def test_ensure_get_signing_from_jwt_called(self, mock_get_signing_key_from_jwt):
        # Mock the public signing.key used in the auth function
        mock_get_signing_key_from_jwt.side_effect = [self.signing]
        token = TestOIDC.generate_token(self.parsed_token, self.private_key)

        self.loop.run_until_complete(self.auth(token))

        token = jwt.encode(self.parsed_token, self.private_key, algorithm="RS256")

        mock_get_signing_key_from_jwt.assert_called_once_with(token)

    @unittest.mock.patch(
        "os2mo_fastapi_utils.auth.oidc.jwt.PyJWKClient.get_signing_key_from_jwt"
    )
    def test_exception_when_aud_in_token_and_audience_is_not_set(
        self, mock_get_signing_key_from_jwt
    ):
        # Mock the public signing.key used in the auth function
        mock_get_signing_key_from_jwt.side_effect = [self.signing]

        # Set aud in token
        self.parsed_token["aud"] = "mo"
        token = TestOIDC.generate_token(self.parsed_token, self.private_key)

        with self.assertRaises(AuthenticationError) as err:
            self.loop.run_until_complete(self.auth(token))
            self.assertTrue(isinstance(err.exception.exc, InvalidAudienceError))

    @unittest.mock.patch(
        "os2mo_fastapi_utils.auth.oidc.jwt.PyJWKClient.get_signing_key_from_jwt"
    )
    def test_token_accepted_when_aud_in_token_and_audience_is_set(
        self, mock_get_signing_key_from_jwt
    ):
        # Mock the public signing.key used in the auth function
        mock_get_signing_key_from_jwt.side_effect = [self.signing]

        # Set aud in token
        self.parsed_token["aud"] = "mo"
        token = TestOIDC.generate_token(self.parsed_token, self.private_key)

        auth = get_auth_dependency(
            host="keycloak",
            port=8080,
            realm="mo",
            token_url_path="service/token",
            token_model=Token,
            audience="mo",
        )

        assert self.loop.run_until_complete(auth(token))

    @unittest.mock.patch(
        "os2mo_fastapi_utils.auth.oidc.jwt.PyJWKClient.get_signing_key_from_jwt"
    )
    def test_token_accepted_when_aud_in_token_and_verify_aud_is_false(
        self, mock_get_signing_key_from_jwt
    ):
        # Mock the public signing.key used in the auth function
        mock_get_signing_key_from_jwt.side_effect = [self.signing]

        # Set aud in token
        self.parsed_token["aud"] = "mo"
        token = TestOIDC.generate_token(self.parsed_token, self.private_key)

        auth = get_auth_dependency(
            host="keycloak",
            port=8080,
            realm="mo",
            token_url_path="service/token",
            token_model=Token,
            verify_audience=False,
        )

        assert self.loop.run_until_complete(auth(token))


class TestAuthError(unittest.TestCase):
    """
    Test that the AuthError exception itself works as expected
    """

    def test_exception_set_correctly(self):
        exc = AuthenticationError(InvalidSignatureError())
        self.assertTrue(InvalidSignatureError, type(exc.exc))

    def test_is_client_side_error_true_for_invalid_token_error(self):
        exc = AuthenticationError(InvalidTokenError())
        self.assertTrue(exc.is_client_side_error())

    def test_is_client_side_error_true_for_http_exception_401(self):
        exc = AuthenticationError(HTTPException(status_code=HTTP_401_UNAUTHORIZED))
        self.assertTrue(exc.is_client_side_error())

    def test_is_client_side_error_false_for_http_exception_500(self):
        exc = AuthenticationError(
            HTTPException(status_code=HTTP_500_INTERNAL_SERVER_ERROR)
        )
        self.assertFalse(exc.is_client_side_error())

    def test_is_client_side_error_false_for_server_error(self):
        exc = AuthenticationError(PyJWTError())
        self.assertFalse(exc.is_client_side_error())


class TestTokenModel(unittest.TestCase):
    def setUp(self) -> None:
        self.roles = {"admin", "owner"}
        self.realm_access = RealmAccess(roles=self.roles)

    def test_azp_mandatory(self):
        with self.assertRaises(ValidationError) as err:
            Token()
        errors = err.exception.errors()[0]

        self.assertEqual(("azp",), errors["loc"])
        self.assertEqual("value_error.missing", errors["type"])

    def test_should_ignore_extra_fields(self):
        token = Token(azp="some-client", extra=0)
        with self.assertRaises(AttributeError):
            token.extra

    def test_should_set_realm_access_to_default_value(self):
        token = Token(azp="some-client")
        self.assertSetEqual(set(), token.realm_access.roles)

    def test_should_set_roles_correctly(self):
        token = Token(azp="some-client", realm_access=self.realm_access)
        self.assertSetEqual(self.roles, token.realm_access.roles)

    def test_invalid_email_address_not_allowed(self):
        with self.assertRaises(ValidationError) as err:
            Token(azp="mo", uuid=uuid4(), email="Invalid email")
        errors = err.exception.errors()[0]

        self.assertEqual(("email",), errors["loc"])
        self.assertEqual("value_error.email", errors["type"])

    def test_token_with_all_fields_set(self):
        uuid = uuid4()
        token = Token(
            azp="mo",
            email="bruce@kung.fu",
            preferred_username="bruce",
            realm_access=self.realm_access,
            uuid=uuid,
        )

        self.assertEqual("mo", token.azp)
        self.assertEqual("bruce@kung.fu", token.email)
        self.assertEqual("bruce", token.preferred_username)
        self.assertSetEqual(self.roles, token.realm_access.roles)
        self.assertEqual(uuid, token.uuid)


tests/test_tracing.py
# SPDX-FileCopyrightText: Magenta ApS
#
# SPDX-License-Identifier: MPL-2.0
from unittest import TestCase

from fastapi import FastAPI
from fastapi.testclient import TestClient
from structlog import get_logger
from structlog.testing import CapturingLoggerFactory

from os2mo_fastapi_utils.tracing import setup_instrumentation
from os2mo_fastapi_utils.tracing import setup_logging


def create_app():
    app = FastAPI()

    def other_function():
        logger = get_logger()
        logger.info("other_function called", r2="d2")

    @app.get("/")
    async def root():
        other_function()
        return {"message": "Hello World"}

    app = setup_instrumentation(app)

    return app


class TracingTests(TestCase):
    def setUp(self):
        self.app = create_app()
        self.client = TestClient(self.app)

    def test_app(self):
        response = self.client.get("/")
        self.assertEqual(response.status_code, 200)
        payload = response.json()
        self.assertEqual(payload, {"message": "Hello World"})

    def test_response_headers(self):
        response = self.client.get("/")
        trace_id = response.headers["X-Trace-ID"]
        span_id = response.headers["X-Span-ID"]
        # Check that we got two hex values back
        self.assertIn("0x", trace_id)
        self.assertIn("0x", span_id)
        int(trace_id, 16)
        int(span_id, 16)

    def test_logs(self):
        cf = CapturingLoggerFactory()
        setup_logging(logger_factory=cf)

        self.client.get("/")
        self.assertEqual(len(cf.logger.calls), 3)

        logmsg = cf.logger.calls[0].kwargs
        self.assertEqual(logmsg["event"], "Request received")
        self.assertIn("0x", logmsg["trace_id"])
        self.assertIn("0x", logmsg["span_id"])
        trace_id = logmsg["trace_id"]
        span_id = logmsg["span_id"]
        self.assertEqual(logmsg["method"], "GET")
        self.assertEqual(logmsg["url"], "/")

        logmsg = cf.logger.calls[1].kwargs
        self.assertEqual(logmsg["event"], "other_function called")
        self.assertEqual(logmsg["trace_id"], trace_id)
        self.assertEqual(logmsg["span_id"], span_id)
        self.assertEqual(logmsg["r2"], "d2")

        logmsg = cf.logger.calls[2].kwargs
        self.assertEqual(logmsg["event"], "Request processed")
        self.assertEqual(logmsg["trace_id"], trace_id)
        self.assertEqual(logmsg["span_id"], span_id)

    def test_x_request_id(self):
        request_id = "Alfa Beta Mogens"

        # Request without x-request-id header
        cf = CapturingLoggerFactory()
        setup_logging(logger_factory=cf)

        self.client.get("/")
        self.assertGreaterEqual(len(cf.logger.calls), 1)

        logmsg = cf.logger.calls[0].kwargs
        self.assertNotIn("request_id", logmsg)

        # Request with x-request-id header
        cf = CapturingLoggerFactory()
        setup_logging(logger_factory=cf)

        self.client.get("/", headers={"X-Request-ID": request_id})
        self.assertGreaterEqual(len(cf.logger.calls), 1)

        logmsg = cf.logger.calls[0].kwargs
        self.assertIn("request_id", logmsg)
        self.assertEqual(logmsg["request_id"], request_id)

        # Request with x-request-id header
        cf = CapturingLoggerFactory()
        setup_logging(logger_factory=cf)

        self.client.get("/", headers={"x-rEqUeSt-Id": request_id})
        self.assertGreaterEqual(len(cf.logger.calls), 1)

        logmsg = cf.logger.calls[0].kwargs
        self.assertIn("request_id", logmsg)
        self.assertEqual(logmsg["request_id"], request_id)


tests/test_pydantic_types.py
# SPDX-FileCopyrightText: Magenta ApS
#
# SPDX-License-Identifier: MPL-2.0
from unittest import TestCase

from pydantic import BaseModel
from pydantic import ValidationError

from os2mo_fastapi_utils.pydantic_types import Domain
from os2mo_fastapi_utils.pydantic_types import Port


class PortModel(BaseModel):
    port: Port


class DomainModel(BaseModel):
    domain: Domain


class TypeTests(TestCase):
    def test_create_port(self):
        port = Port(80)
        self.assertEqual(repr(port), "Port(80)")
        self.assertEqual(str(port), "Port(80)")

        other_port = Port(80)
        self.assertEqual(port, other_port)

        yet_another_port = Port(81)
        self.assertNotEqual(yet_another_port, port)
        self.assertNotEqual(yet_another_port, other_port)

    def test_valid_port(self):
        PortModel(port=80)
        PortModel(port=5432)

    def test_invalid_port(self):
        with self.assertRaises(ValidationError):
            PortModel(port=70000)

        with self.assertRaises(ValidationError):
            PortModel(port="80")

    def test_create_domain(self):
        domain = Domain("example.org")
        self.assertEqual(repr(domain), "Domain('example.org')")
        self.assertEqual(str(domain), "example.org")

        other_domain = Domain("example.org")
        self.assertEqual(domain, other_domain)

        yet_another_domain = Domain("example.com")
        self.assertNotEqual(yet_another_domain, domain)
        self.assertNotEqual(yet_another_domain, other_domain)

    def test_valid_domain(self):
        DomainModel(domain="aa")
        DomainModel(domain="æøå.com")

    def test_invalid_domain(self):
        with self.assertRaises(ValidationError):
            DomainModel(domain=".")

        with self.assertRaises(ValidationError):
            DomainModel(domain=42)


os2mo_fastapi_utils/tracing.py
# SPDX-FileCopyrightText: Magenta ApS
#
# SPDX-License-Identifier: MPL-2.0
from contextlib import contextmanager
from typing import Optional

import structlog
from fastapi import Request
from fastapi import Response
from opentelemetry import trace
from opentelemetry.exporter.jaeger.thrift import JaegerExporter
from opentelemetry.instrumentation.aiohttp_client import AioHttpClientInstrumentor
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.httpx import HTTPXClientInstrumentor
from opentelemetry.instrumentation.requests import RequestsInstrumentor
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from pydantic import BaseSettings
from structlog import get_logger
from structlog.contextvars import bind_contextvars
from structlog.contextvars import clear_contextvars
from structlog.contextvars import merge_contextvars
from structlog.contextvars import unbind_contextvars

from os2mo_fastapi_utils.pydantic_types import Domain
from os2mo_fastapi_utils.pydantic_types import Port


class TracingSettings(BaseSettings):
    jaeger_service: str = "OS2mo_fastapi_utils"
    jaeger_hostname: Optional[Domain] = None
    jaeger_port: Port = Port(6831)


def _get_settings(**overrides) -> TracingSettings:
    return TracingSettings(**overrides)


def setup_instrumentation(app):
    settings: TracingSettings = _get_settings()

    _TRACE_PROVIDER = TracerProvider(
        resource=Resource.create({"service.name": settings.jaeger_service})
    )
    trace.set_tracer_provider(_TRACE_PROVIDER)

    if settings.jaeger_hostname:  # pragma: no cover
        _JAEGER_EXPORTER = JaegerExporter(
            agent_host_name=settings.jaeger_hostname,
            agent_port=settings.jaeger_port,
            udp_split_oversized_batches=True,
        )

        _TRACE_PROVIDER.add_span_processor(BatchSpanProcessor(_JAEGER_EXPORTER))

    AioHttpClientInstrumentor().instrument()
    RequestsInstrumentor().instrument()
    HTTPXClientInstrumentor().instrument()

    # Register logging middleware
    app.middleware("http")(_log_requests_middleware)
    app.middleware("http")(_bind_logger_tracecontext_middleware)

    FastAPIInstrumentor.instrument_app(app)
    return app


@contextmanager
def _bind_context(**kwargs):
    """Context manager to bind logging contextvars."""
    # Filter None values before binding vars
    filtered_kwargs = {
        key: value for (key, value) in kwargs.items() if value is not None
    }

    bind_contextvars(**filtered_kwargs)
    yield
    unbind_contextvars(*filtered_kwargs.keys())


async def _bind_logger_tracecontext_middleware(request: Request, call_next) -> Response:
    """Bind request and tracing variables to logging."""
    # X-Request-ID can be used to assign a user-defined value to all log messages
    # and to the tracing span context.
    request_id = request.headers.get("x-request-id")
    current_span = trace.get_current_span()
    if request_id is not None:
        current_span.set_attribute("request_id", request_id)

    # The Trace-ID and Span-ID from the spancontext will be bound to log messages
    spancontext = current_span.get_span_context()
    trace_id = hex(spancontext.trace_id)
    span_id = hex(spancontext.span_id)

    clear_contextvars()
    with _bind_context(trace_id=trace_id):
        with _bind_context(span_id=span_id):
            with _bind_context(request_id=request_id):
                response = await call_next(request)

    # Set response headers with tracing information
    if request_id is not None:
        response.headers["X-Request-ID"] = request_id
    response.headers["X-Trace-ID"] = trace_id
    response.headers["X-Span-ID"] = span_id

    return response


async def _log_requests_middleware(request: Request, call_next):
    """Log that a request has been received and processed."""
    logger = get_logger()
    logger.debug(
        "Request received",
        method=request.method,
        url=request.url.path,
    )

    response = await call_next(request)

    logger.debug("Request processed")
    return response


def setup_logging(**kwargs):
    """Wrapper around structlog.configure, to add merge_contextvar.

    NOTE: When using this you must yourself register a key-value processor.
    """
    processors = kwargs.get("processors", [])
    if merge_contextvars not in processors:
        # We prepend, as merge_contextvars must come before the key-value processor.
        processors.insert(0, merge_contextvars)
    kwargs["processors"] = processors

    structlog.configure(**kwargs)


os2mo_fastapi_utils/pydantic_types.py
# SPDX-FileCopyrightText: Magenta ApS
#
# SPDX-License-Identifier: MPL-2.0
from typing import Any
from typing import no_type_check
from typing import Optional
from typing import TYPE_CHECKING

from pydantic.errors import UrlHostError
from pydantic.errors import UrlPortError
from pydantic.networks import ascii_domain_regex
from pydantic.networks import int_domain_regex
from pydantic.validators import str_validator

if TYPE_CHECKING:  # pragma: no cover
    from pydantic.typing import CallableGenerator


class Port(int):
    @no_type_check
    def __new__(cls, port: Optional[int]) -> object:
        return int.__new__(cls, port)

    def __init__(self, port: int, *args, **kwargs) -> None:
        int.__init__(port)

    @classmethod
    def __get_validators__(cls) -> "CallableGenerator":
        yield cls.validate

    @classmethod
    def validate(cls, value: Any) -> "Port":
        if not isinstance(value, int):
            raise TypeError("Integer required")

        if value is not None and value > 65_535:
            raise UrlPortError()

        return cls(value)

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}({super().__repr__()})"


class Domain(str):
    strip_whitespace = True

    @no_type_check
    def __new__(cls, domain: Optional[str]) -> object:
        return str.__new__(cls, domain)

    def __init__(self, domain: str, *args, **kwargs) -> None:
        str.__init__(domain)

    @classmethod
    def __get_validators__(cls) -> "CallableGenerator":
        yield cls.validate

    @classmethod
    def validate(cls, value: Any) -> "Domain":
        if not isinstance(value, str):
            raise TypeError("String required")

        value = str_validator(value)
        if cls.strip_whitespace:
            value = value.strip()

        host = value

        is_international = False
        d = ascii_domain_regex().fullmatch(host)
        if d is None:
            d = int_domain_regex().fullmatch(host)
            if d is None:
                raise UrlHostError()
            is_international = True

        tld = d.group("tld")
        if tld is None and not is_international:
            is_international = True

        if is_international:
            host = host.encode("idna").decode("ascii")

        return cls(host)

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}({super().__repr__()})"


os2mo_fastapi_utils/auth/exceptions.py
# SPDX-FileCopyrightText: 2019-2020 Magenta ApS
# SPDX-License-Identifier: MPL-2.0
from fastapi.exceptions import HTTPException
from jwt.exceptions import InvalidTokenError
from pydantic.error_wrappers import ValidationError
from starlette.status import HTTP_401_UNAUTHORIZED


class AuthenticationError(Exception):
    """
    Raised if the user is not authenticated.
    """

    def __init__(self, exc: Exception):
        self.__exc = exc

    @property
    def exc(self) -> Exception:
        return self.__exc

    def is_client_side_error(self) -> bool:
        """
        Return True if the error is a client side error (e.g. an expired
        token) and False otherwise (e.g. if Keycloak is unreachable)
        """
        if isinstance(self.__exc, InvalidTokenError):
            return True
        if isinstance(self.__exc, ValidationError):
            return True
        if (
            isinstance(self.__exc, HTTPException)
            and self.__exc.status_code == HTTP_401_UNAUTHORIZED
        ):
            return True
        return False


os2mo_fastapi_utils/auth/oidc.py
# SPDX-FileCopyrightText: 2019-2020 Magenta ApS
# SPDX-License-Identifier: MPL-2.0
from typing import Any
from typing import List
from typing import Type
from typing import TypeVar
from typing import Union

import jwt.exceptions
from fastapi import Depends
from fastapi import Request
from fastapi.security import OAuth2PasswordBearer
from starlette.responses import JSONResponse
from starlette.status import HTTP_401_UNAUTHORIZED
from starlette.status import HTTP_500_INTERNAL_SERVER_ERROR

from os2mo_fastapi_utils.auth.exceptions import AuthenticationError
from os2mo_fastapi_utils.auth.models import Token

TokenModel = TypeVar("TokenModel", bound=Token)


def get_auth_dependency(
    host: str,
    port: int,
    realm: str,
    token_url_path: str,
    token_model: Type[TokenModel],
    http_schema: str = "http",
    alg: str = "RS256",
    verify_audience: bool = True,
    audience: Union[str, List[str], None] = None,
):
    # URI for obtaining JSON Web Key Set (JWKS), i.e. the public Keycloak key
    JWKS_URI = (
        f"{http_schema}://{host}:{port}"
        f"/auth/realms/{realm}/protocol/openid-connect/certs"
    )

    # JWKS client for fetching and caching JWKS
    jwks_client = jwt.PyJWKClient(JWKS_URI)

    # For getting and parsing the Authorization header
    oauth2_scheme = OAuth2PasswordBearer(tokenUrl=token_url_path)

    async def keycloak_auth(token: str = Depends(oauth2_scheme)) -> TokenModel:
        """
        Ensure the caller has a valid OIDC token, i.e. that the Authorization
        header is set with a valid bearer token.

        :param token: encoded Keycloak token
        :return: selected JSON values from the Keycloak token

        **Example return value**

        .. sourcecode:: json

            {
                "acr": "1",
                "allowed-origins": ["http://localhost:5001"],
                "azp": "mo",
                "email": "bruce@kung.fu",
                "email_verified": false,
                "exp": 1621779689,
                "family_name": "Lee",
                "given_name": "Bruce",
                "iat": 1621779389,
                "iss": "http://localhost:8081/auth/realms/mo",
                "jti": "25dbb58d-b3cb-4880-8b51-8b92ada4528a",
                "name": "Bruce Lee",
                "preferred_username": "bruce",
                "realm_access": {
                    "roles": [
                      "admin"
                  ]
                },
                "scope": "email profile",
                "session_state": "d94f8dc3-d930-49b3-a9dd-9cdc1893b86a",
                "sub": "c420894f-36ba-4cd5-b4f8-1b24bd8c53db",
                "typ": "Bearer",
                "uuid": "99e7b256-7dfa-4ee8-95c6-e3abe82e236a"
            }

        """

        try:
            # Get the public signing key from Keycloak. The JWKS client uses an
            # lru_cache, so it will not make an HTTP request to Keycloak each time
            # get_signing_key_from_jwt() is called.

            signing = jwks_client.get_signing_key_from_jwt(token)

            # The jwt.decode() method raises an exception (e.g.
            # InvalidSignatureError, ExpiredSignatureError,...) in case the OIDC
            # token is invalid. These exceptions will be caught by the
            # auth_exception_handler below which is used by the FastAPI app.

            # The audience verification can be disabled (aud
            # claim in the token) when all services in the stack trust
            # each other
            # (see https://www.keycloak.org/docs/latest/server_admin/index.html#_audience)
            decoded_token: dict = jwt.decode(
                token,
                signing.key,
                algorithms=[alg],
                audience=audience,
                options={"verify_aud": verify_audience},
                leeway=5,
            )

            return token_model.parse_obj(decoded_token)

        except Exception as err:
            raise AuthenticationError(err)

    return keycloak_auth


def get_auth_exception_handler(logger: Any):
    """
    Returns an authentication exception handler to be used by the FastAPI
    app object

    :param logger: any logger for logging auth errors
    """

    def authentication_exception_handler(
        request: Request, err: AuthenticationError
    ) -> JSONResponse:
        if err.is_client_side_error():
            logger.exception("Client side authentication error", exc_info=err)
            return JSONResponse(
                status_code=HTTP_401_UNAUTHORIZED,
                content={"status": "Unauthorized", "msg": str(err.exc)},
            )

        logger.exception("Problem communicating with the Keycloak server", exc_info=err)

        return JSONResponse(
            status_code=HTTP_500_INTERNAL_SERVER_ERROR,
            content={"msg": "A server side authentication error occurred"},
        )

    return authentication_exception_handler
